components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/DefaultAPIHandler.java

```diff
@@ -0,0 +1,71 @@
+package org.wso2.carbon.apimgt.gateway.handlers;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.synapse.AbstractSynapseHandler;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.api.ApiUtils;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.rest.RESTConstants;
+import org.wso2.carbon.apimgt.gateway.utils.GatewayUtils;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+import org.wso2.carbon.apimgt.keymgt.SubscriptionDataHolder;
+import org.wso2.carbon.apimgt.keymgt.model.SubscriptionDataStore;
+import org.wso2.carbon.apimgt.keymgt.model.entity.API;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class DefaultAPIHandler extends AbstractSynapseHandler {
+    @Override
+    public boolean handleRequestInFlow(MessageContext messageContext) {
+        org.apache.axis2.context.MessageContext axis2MessageContext =
+                ((Axis2MessageContext) messageContext).getAxis2MessageContext();
+        SubscriptionDataStore tenantSubscriptionStore =
+                SubscriptionDataHolder.getInstance().getTenantSubscriptionStore(GatewayUtils.getTenantDomain());
+        if (tenantSubscriptionStore != null) {
+            Map<String, API> contextAPIMap = tenantSubscriptionStore.getAllAPIsByContextList();
+            String path = ApiUtils.getFullRequestPath(messageContext);
+            List<API> selectedAPIS = new ArrayList<>();
+            if (contextAPIMap != null) {
+                contextAPIMap.forEach((context, api) -> {
+                    if (ApiUtils.matchApiPath(path, context)) {
+                        selectedAPIS.add(api);
+                    }
+                });
+                if (selectedAPIS.size() > 0) {
```

Comment:
we need to loop or move this inside foreach , to get the appropriate api

Reply:
we first need to select all available apis for match.
Then we have to check if api count was > 0 then do the necessary

Reply:
instead, when selecting the matching apis, we can see if that is the default version and if so, we don't need to find the rest of the apis.

Reply:
use-case
API1 - /api1/1.0.0
Default API /api1 -> /api1/1.0.0

When a request comes as /api1/1.0.0 we can't stop at the default version since the exact match was /api1/1.0.0 version api. Therefore we first have to select elected APIs and then iterate them to get most suitable one.

@malinthaprasan  wdyt ?.

Reply:
Yes we need to select the most suitable API. But have a small question

If there are conflicting paths like:
API1 - api/1.0  (with default version)
API2 - api/ctx/1.0 (also with default version)

if someone sends GET /api/ctx/resource will this correctly route to the API2?